src/middlewared/middlewared/main.py

```diff
@@ -1328,6 +1334,36 @@ def pipe(self, buffered=False):
         """
         return Pipe(self, buffered)
 
+    def _raise_if_rate_limit_exceeded(self, name, ip):
+        """Should the method named `name` be rate limited? If so, raise
+        an exception accordingly.
+
+        NOTE: This is automatically applied to all endpoints that are marked
+        with the `@no_auth_required` decorator. This is needed, for example, to
+        prevent anyone that knows the IP address of the TrueNAS from hammering
+        these endpoints endlessly causing the SoftSemaphoreLimit to be hit by
+        a single unauthenticated user.
+        """
+        if ip is None:
+            return
+
+        try:
+            info = self._rate_limit_info[f'{name}_{ip}']
+        except KeyError:
+            return
+        else:
+            with info['lock']:
+                now = time.monotonic()
+                if info['max_period'] - (now - info['last_reset']) <= 0:
+                    # time window elapsed, so time to reset
+                    info['num_times_called'] = 0
+                    info['last_reset'] = now
+
+                # always increment
+                info['num_times_called'] += 1
+                if info['num_times_called'] > info['max_calls']:
+                    raise CallError('rate limit exceeded')
```

Comment:
Maybe set an errno like EBUSY.

Reply:
Would it be a good idea to put a sleep in before raising the CallError to make a client back off a bit if there's a bug where application is in tight loop calling the endpoint? I've seen UI do this before.